CRM/Mailing/BAO/Recipients.php

```diff
@@ -132,4 +132,39 @@ public static function reassign($sourceMailingId, $to) {
     }
   }
 
+  /**
+  * Function to delete from mailing_recipients in chunks to avoid locking 
+  * up civicrm_contacts and civicrm_emails. Returns the total count
+  * of deleted rows.
+  *
+  * @param int $mailingID
+  *
+  * @return int
+  */
+  public static function clearRecipients($mailingID) {
+
+    // print("Let's go \n");
+
+    $deleted = true;
+    $chunk_size = 10000;
+    $total = 0;
+ 
+    while ($deleted) {
+      $recipients = new CRM_Mailing_DAO_Recipients();
+      $recipients->whereAdd("mailing_id = $mailingID");
+      $recipients->limit($chunk_size);
+
+      $rows = $recipients->delete(true);
+      $total += $rows;
+      $deleted = ($rows > 0);
+
+      // print("Rows $rows\n");
+
+      # COMMIT to release row locks on related tables
+      CRM_Core_DAO::executeQuery("COMMIT");
```

Comment:
Do you understand correctly that this is to break any potential transaction started up-stack? We give up any hope of merging this upstream, right?

Reply:
Yep !

That's the only way to release the row locks.